roles/download/tasks/main.yml

```diff
@@ -16,6 +16,43 @@
     - download
     - upload
 
+# The image_info_command depends on the Container Runtime and will output something like the following:
+# nginx:1.15,gcr.io/google-containers/kube-proxy:v1.14.1,gcr.io/google-containers/kube-proxy@sha256:44af2833c6cbd9a7fc2e9d2f5244a39dfd2e31ad91bf9d4b7d810678db738ee9,gcr.io/google-containers/kube-apiserver:v1.14.1,etc...
+- name: Download | Generate a list of information about the images on a node  # noqa command-instead-of-shell - image_info_command contains a pipe, therefore requiring shell
+  shell: "{{ image_info_command }}"
+  register: node_images
+  changed_when: false
+  check_mode: no
+  when: not download_always_pull
+
+- name: Show node images
+  debug:
+    msg: "{{ node_images.stdout }}"
+  when: not download_always_pull
+
+# convert all items in downloads which have the key container set to true
+# and check if the container exists in node_images.stdout.split(',')
+# if yes, set the enabled key to false
+# TODO: Some images are not cached on all nodes, which leads to repeated downloads... I still don't know why this is the case.
+# TODO: Adapt to also support pull by sha256
+- name: Download | Check if the container image is already present
+  set_fact:
+    downloads: "{{ downloads | combine({item.key: item.value | combine({'enabled': false})}) }}"
+  loop: "{{ downloads | dict2items }}"
```

Comment:
I would avoid adding yet another loop.
I agree that "patching" `downloads` at runtime is a good idea, but using filter in the `include_tasks` loop below would be better.
If the ansible filters are not up to the tasks, use `json_query`, it's quite powerful (but can be headache inducing if abused :laughing: )

Reply:
Thanks I didn't want to jump on the loop at first (I was a little put off by the loop 😄).
I have moved the logic to the loop task. It's not a `json_query` but it should still save us the extra loop.

What do you think?